nomad/state/state_store.go

```diff
@@ -6959,3 +6959,36 @@ func (s *StateStore) GetActiveRootKeyMeta(ws memdb.WatchSet) (*structs.RootKeyMe
 	}
 	return nil, nil
 }
+
+// IsRootKeyMetaInUse determines whether a key has been used to sign a workload
+// identity for a live allocation or encrypt any variables
+func (s *StateStore) IsRootKeyMetaInUse(keyID string) (bool, error) {
```

Comment:
This made me a little nervous at first glance because it could trend toward quadratic over time if there's at-least-one alloc for every key that's ever existed...

...however at the default key rotation rate of 30 days even in the maximally pathological case of 1-alloc-for-every-key-ever, that's a whopping 120 iterations over all allocations after *a decade.*

If we ever become concerned with the performance here we can always add ref counting to keys and update it when allocs are created or updated (or even just wait to decrement when the allocs are gc'd if we want to be really lazy). I'm guessing we'll never need to do that.

Making a compound index that only includes non-terminal allocs would be an easier optimization if not quite as effective.

Reply:
Yeah, there are a lot of cases where I want a cheap `Count()` method in memdb. 😀 I can do that compound index though: it reduces the memory usage for the index a bit for clusters with lots of short-lived allocs, and reduces the number of iterations we have to do here at a tiny cost on `txn.Insert`. Seems worth it.